### PR TITLE
BO Migration Shop parameters > Contact > Stores - Add Store entity

### PR DESCRIPTION
**Pull request by tom-combet**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      |  Add Store entity
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Related to #10624.
| How to test?      | N/A
| Possible impacts? | None.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27388)

### CODE REVIEW

File: src/Core/ConstraintValidator/TypedRegexValidator.php [hunk 1]
```diff
@@ -141,6 +141,8 @@ private function getPattern($type)
                 return '/^[a-zA-Z0-9_-]+$/';
             case TypedRegex::TYPE_URL:
                 return $this->characterCleaner->cleanNonUnicodeSupport('/^[~:#,$%&_=\(\)\.\? \+\-@\/a-zA-Z0-9\pL\pS-]+$/u');
+            case TypedRegex::TYPE_COORDINATE:
+                return '/^\-?[0-9]{1,8}\.[0-9]{1,8}$/s';
```
Comment (Progi1984) [inline src/Core/ConstraintValidator/TypedRegexValidator.php:145]: Could you add an unit test `tests/Unit/Core/ConstraintValidator/TypedRegexValidatorTest.php` ?

File: src/PrestaShopBundle/Entity/Repository/StoreLangRepository.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShopBundle\Entity\Repository;
```
Comment (Progi1984) [inline src/PrestaShopBundle/Entity/Repository/StoreLangRepository.php]:
```suggestion
 */

declare(strict_types=1):

namespace PrestaShopBundle\Entity\Repository;
```

File: src/PrestaShopBundle/Entity/Repository/StoreRepository.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShopBundle\Entity\Repository;
```
Comment (Progi1984) [inline src/PrestaShopBundle/Entity/Repository/StoreRepository.php]:
```suggestion
 */

declare(strict_types=1):

namespace PrestaShopBundle\Entity\Repository;
```

File: src/PrestaShopBundle/Entity/Store.php [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShopBundle\Entity;
```
Comment (Progi1984) [inline src/PrestaShopBundle/Entity/Store.php]:
```suggestion
 */

declare(strict_types=1):

namespace PrestaShopBundle\Entity;
```

File: src/PrestaShopBundle/Entity/Store.php [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+
+        return $this;
+    }
+
+    public function getShops(): Collection
+    {
+        return $this->shops;
+    }
+
+    public function addStoreLang(StoreLang $storeLang)
```
Comment (Progi1984) [inline src/PrestaShopBundle/Entity/Store.php]:
```suggestion
    public function addStoreLang(StoreLang $storeLang): self
```

File: src/PrestaShopBundle/Entity/Store.php [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+        $this->storeLangs[] = $storeLang;
+
+        $storeLang->setStore($this);
+
+        return $this;
+    }
+
+    public function removeStoreLang(StoreLang $storeLang)
+    {
+        $this->storeLangs->removeElement($storeLang);
```
Comment (Progi1984) [inline src/PrestaShopBundle/Entity/Store.php]:
```suggestion
    public function removeStoreLang(StoreLang $storeLang): self
    {
        $this->storeLangs->removeElement($storeLang);
        
        return $this;
```

File: src/PrestaShopBundle/Entity/Store.php [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+    }
+
+    public function getStoreLangs()
```
Comment (Progi1984) [inline src/PrestaShopBundle/Entity/Store.php]:
```suggestion
    public function getStoreLangs(): ArrayCollection
```

File: src/PrestaShopBundle/Entity/Store.php [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+    }
+
+    public function removeShop(Shop $shop): self
+    {
+        $this->shops->removeElement($shop);
+
+        return $this;
+    }
+
+    public function getShops(): Collection
```
Comment (Progi1984) [inline src/PrestaShopBundle/Entity/Store.php]:
```suggestion
    public function getShops(): ArrayCollection
```

File: src/PrestaShopBundle/Entity/StoreLang.php [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShopBundle\Entity;
```
Comment (Progi1984) [inline src/PrestaShopBundle/Entity/StoreLang.php]:
```suggestion
 */

declare(strict_types=1):

namespace PrestaShopBundle\Entity;
```

**Review by Progi1984 — COMMENTED** [review]
@Kaudaj Some feedbacks

File: src/PrestaShopBundle/Entity/Repository/StoreLangRepository.php [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShopBundle\Entity\Repository;
+
+use Doctrine\ORM\EntityRepository;
+
+class StoreLangRepository extends EntityRepository
```
Comment (matks) [inline src/PrestaShopBundle/Entity/Repository/StoreLangRepository.php]: Hello as the PR does not contain the code that will make use of this class I cannot see if it's a good implementation or not.

I know you want to split the PR into small bits but if I don't have the context I cannot provide feedback about whether or not this code is valid.

File: src/PrestaShopBundle/Entity/Store.php [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+
+namespace PrestaShopBundle\Entity;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\ORM\Mapping as ORM;
+use PrestaShop\PrestaShop\Core\ConstraintValidator\Constraints as PrestaShopAssert;
+use PrestaShopBundle\Entity\Repository\StoreRepository;
+use Symfony\Component\Validator\Constraints as Assert;
+
+/**
```
Comment (matks) [inline src/PrestaShopBundle/Entity/Store.php:35]: Hello as the PR does not contain the code that will make use of this class I cannot see if it's a good implementation or not.

I know you want to split the PR into small bits but if I don't have the context I cannot provide feedback about whether or not this code is valid.

However so far we have not used Doctrine entities for back-office listing pages. If you look at other back-office listing pages migrated to Symfony we do not use Doctrine entities. So this looks like a wrong path.

File: tests/Unit/Core/ConstraintValidator/TypedRegexValidatorTest.php [hunk 12]
```diff
@@ -270,55 +270,21 @@ public function testItFailsForIsbnTypeWhenInvalidCharacterGiven(string $invalidC
         $this->assertViolationIsRaised(new TypedRegex('isbn'), $invalidChar);
     }
 
-    /**
```
Comment (Progi1984) [inline tests/Unit/Core/ConstraintValidator/TypedRegexValidatorTest.php]: I believe you have removed a lot of unit tests.

File: src/PrestaShopBundle/Entity/Store.php (hunk 11 above)
Comment (tom-combet) [reply src/PrestaShopBundle/Entity/Store.php:35]: Thought Doctrine migration was topical, my bad. I'll finish it and put it aside until then.

File: tests/Unit/Core/ConstraintValidator/TypedRegexValidatorTest.php (hunk 12 above)
Comment (tom-combet) [reply tests/Unit/Core/ConstraintValidator/TypedRegexValidatorTest.php]: Still don't know what happened. Should be good now! 😅

File: src/PrestaShopBundle/Entity/Store.php [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+     */
+    private $countryId;
+
+    /**
+     * @ORM\Column(name="id_state", type="integer", nullable=true, options={"unsigned"=true})
+     */
+    private $stateId;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("post_code")
```
Comment (atomiix) [inline src/PrestaShopBundle/Entity/Store.php]: An entity should always be valid: https://sarvendev.com/2018/01/entity-should-always-be-valid/
```suggestion
```

File: src/PrestaShopBundle/Entity/Store.php [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+     * @ORM\Column(type="string", length=12)
+     */
+    private $postcode;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("city_name")
```
Comment (atomiix) [inline src/PrestaShopBundle/Entity/Store.php]:
```suggestion
```

File: src/PrestaShopBundle/Entity/Store.php [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+     * @ORM\Column(type="string", length=64)
+     */
+    private $city;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("coordinate")
```
Comment (atomiix) [inline src/PrestaShopBundle/Entity/Store.php]:
```suggestion
```

File: src/PrestaShopBundle/Entity/Store.php [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+     * @ORM\Column(type="float", nullable=true)
+     */
+    private $latitude;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("coordinate")
```
Comment (atomiix) [inline src/PrestaShopBundle/Entity/Store.php]:
```suggestion
```

File: src/PrestaShopBundle/Entity/Store.php [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+     * @ORM\Column(type="float", nullable=true)
+     */
+    private $longitude;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("phone_number")
```
Comment (atomiix) [inline src/PrestaShopBundle/Entity/Store.php]:
```suggestion
```

File: src/PrestaShopBundle/Entity/Store.php [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+     * @ORM\Column(type="string", length=16, nullable=true)
+     */
+    private $phone;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("phone_number")
```
Comment (atomiix) [inline src/PrestaShopBundle/Entity/Store.php]:
```suggestion
```

File: src/PrestaShopBundle/Entity/StoreLang.php [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+    /**
+     * @var Lang
+     * @ORM\Id
+     * @ORM\ManyToOne(targetEntity="Lang")
+     * @ORM\JoinColumn(name="id_lang", referencedColumnName="id_lang", nullable=false, onDelete="CASCADE")
+     */
+    private $lang;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("generic_name")
```
Comment (atomiix) [inline src/PrestaShopBundle/Entity/StoreLang.php]:
```suggestion
```

File: src/PrestaShopBundle/Entity/StoreLang.php [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+     * @ORM\Column(type="string", length=255)
+     */
+    private $name;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("address")
```
Comment (atomiix) [inline src/PrestaShopBundle/Entity/StoreLang.php]:
```suggestion
```

File: src/PrestaShopBundle/Entity/StoreLang.php [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+     * @ORM\Column(type="string", length=255)
+     */
+    private $address1;
+
+    /**
+     * @PrestaShopAssert\TypedRegex("address")
```
Comment (atomiix) [inline src/PrestaShopBundle/Entity/StoreLang.php]:
```suggestion
```

File: src/PrestaShopBundle/Entity/StoreLang.php [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+    private $address2;
+
+    /**
+     * @Assert\Json
+     * @ORM\Column(type="text", nullable=true)
+     */
+    private $hours;
+
+    /**
+     * @PrestaShopAssert\CleanHtml
```
Comment (atomiix) [inline src/PrestaShopBundle/Entity/StoreLang.php]:
```suggestion
```

**Review by atomiix — CHANGES_REQUESTED** [review]
Entities should always be valid, the data validation should be done before.
Also, I think that `StoreLangRepository` & `StoreRepository` classes can be removed as they do not contain new methods.

**Comment by tom-combet** [comment]
@atomiix Ah I maybe misunderstood [the doc](https://devdocs.prestashop.com/1.7/development/architecture/migration-guide/validation/#validation), I thought validation was necessary anyway. I'll fix that.

**Review by Progi1984 — CHANGES_REQUESTED** [review]
@Kaudaj Could you remove entities ?

**Comment by tom-combet** [comment]
@Progi1984 Uh, why? This is the whole purpose of this PR... 😆
Can't I leave the PR on hold until the migration integrates the entities? Otherwise, I close it but I don't see the point of just removing the code.

**Comment by Progi1984** [comment]
@tom-combet The migration to entities will begin not in days or weeks,  more like in months or years. I think you can close this PR and keep your branch.

**Pull request closed without merging**